direct/io-bulkfs-parquet/src/main/java/cz/o2/proxima/direct/bulk/fs/parquet/StreamElementMaterializer.java

```diff
@@ -0,0 +1,372 @@
+    this.schema = schema;
+    this.entity = entity;
+    this.attributeNamesPrefix = attributeNamesPrefix;
+    this.root = new ParquetColumnRecordConverter(schema, null, null);
+  }
+
+  @Override
+  public StreamElement getCurrentRecord() {
+
+    Map<String, Object> record = recordToMap(schema, root.getCurrentRecord(), new HashMap<>());
+
+    final String key = getColumnFromRow(ParquetFileFormat.PARQUET_COLUMN_NAME_KEY, record);
+    final String operation =
+        getColumnFromRow(ParquetFileFormat.PARQUET_COLUMN_NAME_OPERATION, record);
+    final String attributeName =
+        getColumnFromRow(ParquetFileFormat.PARQUET_COLUMN_NAME_ATTRIBUTE_PREFIX, record);
+    Optional<AttributeDescriptor<Object>> attribute = entity.findAttribute(attributeName);
+    if (!attribute.isPresent()) {
+      // current attribute is not in entity -> skip
+      log.info(
+          "Skipping attribute [{}] which is not in current entity [{}].",
+          attributeName,
+          entity.getName());
+      return null;
+    }
+    final String uuid = getColumnFromRow(ParquetFileFormat.PARQUET_COLUMN_NAME_UUID, record);
+    final long timestamp =
+        getColumnFromRow(ParquetFileFormat.PARQUET_COLUMN_NAME_TIMESTAMP, record);
+    switch (Operation.of(operation)) {
+      case DELETE:
+        return StreamElement.delete(entity, attribute.get(), uuid, key, attributeName, timestamp);
+      case DELETE_WILDCARD:
+        return StreamElement.deleteWildcard(entity, attribute.get(), uuid, key, timestamp);
+      case UPSERT:
+        return StreamElement.upsert(
+            entity,
+            attribute.get(),
+            uuid,
+            key,
+            attributeName,
+            timestamp,
+            getValueFromCurrentRowData(attribute.get(), record));
+      default:
+        throw new RecordMaterializationException("Unknown operation " + operation);
+    }
+  }
+
+  private byte[] getValueFromCurrentRowData(
+      AttributeDescriptor<?> attribute, Map<String, Object> record) {
+    final String storedAttributeName = attributeNamesPrefix + attribute.toAttributePrefix(false);
+    @SuppressWarnings("unchecked")
+    final ValueSerializer<Object> valueSerializer =
+        (ValueSerializer<Object>) attribute.getValueSerializer();
+    final AttributeValueAccessor<Object, Object> valueAccessor = valueSerializer.getValueAccessor();
+    final SchemaTypeDescriptor<Object> attributeSchema = valueSerializer.getValueSchemaDescriptor();
+    Object attributeValue;
+    if (attributeSchema.isStructureType()) {
+      attributeValue = record.getOrDefault(storedAttributeName, new HashMap<>());
+      @SuppressWarnings("unchecked")
+      final StructureValue cast = StructureValue.of((Map<String, Object>) attributeValue);
+      return valueSerializer.serialize(valueAccessor.createFrom(cast));
+    } else {
+      attributeValue = record.getOrDefault(storedAttributeName, new Object());
```

Comment:
can you please explain this default?

Reply:
Added note into code:
```
There is nothing written into parquet file when we are writing empty element. In that case we need to delegate creation of empty element to serializer. This is reason for that record.getOrDefault. 
```
Hopefully it is more clear now.